src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -66,14 +70,19 @@ public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationP
             var inMemSession = (InMemorySynchronizedStorageSession) session;
             inMemSession.Enlist(() =>
             {
-                if (correlationProperty != SagaCorrelationProperty.None)
+                var lockenTokenKey = $"{sagaData.GetType().FullName}.{(correlationProperty == null ? "None" : correlationProperty.Name)}.{(correlationProperty == null ? "None" : correlationProperty.Value)}";
+                var lockToken = lockers.GetOrAdd(lockenTokenKey, key => new object());
+                lock (lockToken)
                 {
-                    ValidateUniqueProperties(correlationProperty, sagaData);
+                    if (correlationProperty != SagaCorrelationProperty.None)
+                    {
+                        ValidateUniqueProperties(correlationProperty, sagaData);
+                    }
+
+                    data.AddOrUpdate(sagaData.Id,
+                        id => new VersionedSagaEntity(sagaData, lockenTokenKey),
+                        (id, original) => new VersionedSagaEntity(sagaData, lockenTokenKey, original)); // we can never end up here.
```

Comment:
While I don't care for the `AddOrUpdate` calls where we "know" it will only ever be one path, it does seem out of scope for this PR to make changes here. Perhaps as some future larger update to the persistence code.